File: behavioral/chain_of_responsibility/http_request_middleware/python/http_request_middleware.py
from abc import ABC, abstractmethod
from typing import Any, Dict, Optional

class Handler(ABC):
    """
    Abstract Handler class defines the interface for handling requests and
    linking handlers.
    """
    _next_handler: Optional['Handler'] = None

    def set_next(self, handler: 'Handler') -> 'Handler':
        """Sets the next handler in the chain."""
        self._next_handler = handler
        return handler # Allows chaining

    @abstractmethod
    def handle(self, request: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """
        Handles the request. Implementation should call self._next_handler.handle(request)
        if processing should continue, otherwise return a result or None to stop.
        """
        # Default implementation if the chain ends without a specific final handler returning a value
        if self._next_handler:
             return self._next_handler.handle(request)
        return request # Or None, depending on desired end-of-chain default


class LoggingMiddleware(Handler):
    """Logs the incoming request."""
    def handle(self, request: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        print(f"[Log] Received request: {request.get('path', 'N/A')}")
        # Explicitly call the next handler if it exists
        if self._next_handler:
            return self._next_handler.handle(request)
        # If no next handler, return the request or a default response
        return request

class AuthenticationMiddleware(Handler):
    """Checks for authentication credentials."""
    def handle(self, request: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        print("[Auth] Checking authentication...")
        headers = request.get('headers', {})
        if headers.get('Authorization') == 'valid_token':
            print("[Auth] Authentication successful.")
            request['user'] = {'id': 123, 'role': 'admin'}
            # Explicitly call the next handler if it exists
            if self._next_handler:
                return self._next_handler.handle(request)
            # If no next handler, return the (modified) request
            return request
        else:
            print("[Auth] Authentication failed. Aborting request.")
            # Stop the chain by returning the error response directly
            return {'status_code': 401, 'body': 'Unauthorized'}

class AuthorizationMiddleware(Handler):
    """Checks if the authenticated user has the required permissions."""
    def handle(self, request: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        print("[ACL] Checking authorization...")
        user = request.get('user')
        required_role = request.get('required_role')

        if not user:
            print("[ACL] No user found in request (likely due to failed auth). Aborting.")
            # Stop the chain
            return {'status_code': 401, 'body': 'Authentication Required'}

        proceed = False
        if required_role and user.get('role') == required_role:
            print(f"[ACL] Authorization successful (Role: {user.get('role')}).")
            proceed = True
        elif not required_role:
             print("[ACL] No specific role required. Proceeding.")
             proceed = True

        if proceed:
             # Explicitly call the next handler if it exists
            if self._next_handler:
                 return self._next_handler.handle(request)
            # If no next handler, return the request
            return request
        else:
            # This case handles: required_role exists, but user role doesn't match
            print(f"[ACL] Authorization failed. User role '{user.get('role')}' does not match required role '{required_role}'. Aborting.")
            # Stop the chain
            return {'status_code': 403, 'body': 'Forbidden'}


class FinalHandler(Handler):
    """The final processing step if the request passes all middleware."""
    def handle(self, request: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        print("[Final] Processing request...")
        # Simulate actual request processing
        user_id = request.get('user', {}).get('id', 'anonymous')
        body = f"Successfully processed request for path: {request.get('path', 'N/A')} for user {user_id}"
        print("[Final] Request processed.")
        # This is the end of the chain, return the final response
        return {'status_code': 200, 'body': body}

# Helper function (no changes needed)
def make_request(path: str, token: Optional[str] = None, requires_role: Optional[str] = None) -> Dict[str, Any]:
    request: Dict[str, Any] = {'path': path, 'headers': {}}
    if token:
        request['headers']['Authorization'] = token
    if requires_role:
        request['required_role'] = requires_role
    return request

File: behavioral/chain_of_responsibility/http_request_middleware/typescript/http_request_middleware.ts
// Define types (no changes needed)
export type HttpRequest = {
  path: string;
  headers: { [key: string]: string };
  user?: { id: number; role: string };
  requiredRole?: string;
};

export type HttpResponse = {
  statusCode: number;
  body: string;
};

// Interface for all middleware handlers
export interface Middleware {
  setNext(middleware: Middleware): Middleware;
  handle(request: HttpRequest): Promise<HttpResponse>;
}

// Abstract class manages the link, defines the handle signature
export abstract class AbstractMiddleware implements Middleware {
  protected nextMiddleware: Middleware | null = null; // Changed to protected for clarity

  public setNext(middleware: Middleware): Middleware {
    this.nextMiddleware = middleware;
    return middleware;
  }

  // Define the abstract method signature that subclasses must implement
  abstract handle(request: HttpRequest): Promise<HttpResponse>;
}

// Concrete Middleware Implementations

export class LoggingMiddleware extends AbstractMiddleware {
  public async handle(request: HttpRequest): Promise<HttpResponse> {
    console.log(`[Log] Received request: ${request.path}`);
    // Explicitly call the next handler's handle method if it exists
    if (this.nextMiddleware) {
      return await this.nextMiddleware.handle(request);
    }
    // End of chain if this were the last handler
    return {
      statusCode: 200,
      body: "Request logged but not processed further",
    };
  }
}

export class AuthenticationMiddleware extends AbstractMiddleware {
  public async handle(request: HttpRequest): Promise<HttpResponse> {
    console.log("[Auth] Checking authentication...");
    const token = request.headers["authorization"];

    if (token === "valid_token") {
      console.log("[Auth] Authentication successful.");
      request.user = { id: 456, role: "admin" }; // Mutate request
      // Explicitly call the next handler's handle method if it exists
      if (this.nextMiddleware) {
        return await this.nextMiddleware.handle(request);
      }
      // Auth passed, but end of chain
      return {
        statusCode: 200,
        body: "Authentication successful but not processed further",
      };
    } else {
      console.log("[Auth] Authentication failed. Aborting request.");
      // Stop the chain by returning the response directly
      return { statusCode: 401, body: "Unauthorized" };
    }
  }
}

export class AuthorizationMiddleware extends AbstractMiddleware {
  public async handle(request: HttpRequest): Promise<HttpResponse> {
    console.log("[ACL] Checking authorization...");
    const user = request.user;
    const requiredRole = request.requiredRole;

    if (!user) {
      console.log(
        "[ACL] No user found in request (likely due to failed auth). Aborting."
      );
      // Stop the chain
      return { statusCode: 401, body: "Authentication Required" };
    }

    let proceed = false;
    if (requiredRole && user.role === requiredRole) {
      console.log(`[ACL] Authorization successful (Role: ${user.role}).`);
      proceed = true;
    } else if (!requiredRole) {
      console.log("[ACL] No specific role required. Proceeding.");
      proceed = true;
    }

    if (proceed) {
      // Explicitly call the next handler's handle method if it exists
      if (this.nextMiddleware) {
        return await this.nextMiddleware.handle(request);
      }
      // Auth passed, but end of chain
      return {
        statusCode: 200,
        body: "Authorization successful but not processed further",
      };
    } else {
      // This case handles: required_role exists, but user role doesn't match
      console.log(
        `[ACL] Authorization failed. User role '${user.role}' does not match required role '${requiredRole}'. Aborting.`
      );
      // Stop the chain
      return { statusCode: 403, body: "Forbidden" };
    }
  }
}

export class FinalHandler extends AbstractMiddleware {
  // Implement the abstract handle method
  public async handle(request: HttpRequest): Promise<HttpResponse> {
    console.log("[Final] Processing request...");
    const userId = request.user?.id ?? "anonymous"; // Correctly access potential user ID
    const body = `Successfully processed request for path: ${request.path} for user ${userId}`;
    console.log("[Final] Request processed.");
    // This is the end of the chain, return the final response
    return { statusCode: 200, body: body };
  }

  // Override setNext to prevent chaining further (good practice)
  public setNext(middleware: Middleware): Middleware {
    throw new Error("FinalHandler cannot have a next handler.");
  }
}

// Helper function (no changes needed)
export function makeHttpRequest(
  path: string,
  token?: string,
  requiresRole?: string
): HttpRequest {
  const request: HttpRequest = {
    path: path,
    headers: {},
  };
  if (token) {
    request.headers["authorization"] = token;
  }
  if (requiresRole) {
    request.requiredRole = requiresRole;
  }
  return request;
}
